sycl/unittests/scheduler/InOrderQueueDeps.cpp

```diff
@@ -0,0 +1,129 @@
+  *event = reinterpret_cast<pi_event>(1);
+  return PI_SUCCESS;
+}
+
+static pi_result redefinedEnqueueMemUnmap(pi_queue command_queue, pi_mem memobj,
+                                          void *mapped_ptr,
+                                          pi_uint32 num_events_in_wait_list,
+                                          const pi_event *event_wait_list,
+                                          pi_event *event) {
+  EXPECT_EQ(num_events_in_wait_list, 0u);
+  *event = reinterpret_cast<pi_event>(1);
+  return PI_SUCCESS;
+}
+
+static pi_result redefinedEventsWait(pi_uint32 num_events,
+                                     const pi_event *event_list) {
+  return PI_SUCCESS;
+}
+
+pi_result redefinedEventRelease(pi_event event) { return PI_SUCCESS; }
+
+TEST_F(SchedulerTest, InOrderQueueDeps) {
+  default_selector Selector;
+  platform Plt{default_selector()};
+  if (Plt.is_host()) {
+    std::cout << "Not run due to host-only environment\n";
+    return;
+  }
+
+  unittest::PiMock Mock{Plt};
+  Mock.redefine<detail::PiApiKind::piMemBufferCreate>(redefinedMemBufferCreate);
+  Mock.redefine<detail::PiApiKind::piMemRelease>(redefinedMemRelease);
+  Mock.redefine<detail::PiApiKind::piEnqueueMemBufferReadRect>(
+      redefinedEnqueueMemBufferReadRect);
+  Mock.redefine<detail::PiApiKind::piEnqueueMemBufferWriteRect>(
+      redefinedEnqueueMemBufferWriteRect);
+  Mock.redefine<detail::PiApiKind::piEnqueueMemBufferMap>(
+      redefinedEnqueueMemBufferMap);
+  Mock.redefine<detail::PiApiKind::piEnqueueMemUnmap>(redefinedEnqueueMemUnmap);
+  Mock.redefine<detail::PiApiKind::piEventsWait>(redefinedEventsWait);
+  Mock.redefine<detail::PiApiKind::piEventRelease>(redefinedEventRelease);
+
+  context Ctx{Plt};
+  queue InOrderQueue{Ctx, Selector, property::queue::in_order()};
+  cl::sycl::detail::QueueImplPtr InOrderQueueImpl =
+      detail::getSyclObjImpl(InOrderQueue);
+
+  device HostDevice;
+  std::shared_ptr<detail::queue_impl> DefaultHostQueue{
+      new detail::queue_impl(detail::getSyclObjImpl(HostDevice), {}, {})};
+
+  MockScheduler MS;
+
+  int val;
+  buffer<int, 1> Buf(&val, range<1>(1));
+  detail::Requirement Req = getMockRequirement(Buf);
+
+  detail::MemObjRecord *Record =
+      MS.getOrInsertMemObjRecord(InOrderQueueImpl, &Req);
+  MS.getOrCreateAllocaForReq(Record, &Req, InOrderQueueImpl);
+  MS.getOrCreateAllocaForReq(Record, &Req, DefaultHostQueue);
+
+  // Check that sequential memory movements submitted to the same in-order
+  // queue do not depend on each other.
+  detail::Command *Cmd = MS.insertMemoryMove(Record, &Req, DefaultHostQueue);
+  detail::EnqueueResultT Res;
+  MockScheduler::enqueueCommand(Cmd, Res, detail::NON_BLOCKING);
+  Cmd = MS.insertMemoryMove(Record, &Req, InOrderQueueImpl);
+  MockScheduler::enqueueCommand(Cmd, Res, detail::NON_BLOCKING);
+  Cmd = MS.insertMemoryMove(Record, &Req, DefaultHostQueue);
+  MockScheduler::enqueueCommand(Cmd, Res, detail::NON_BLOCKING);
```

Comment:
The test only employs memory-copy commands, isn't it? Should the test be augmented with arbitrary different type of command?

Reply:
Haven't found a good way to verify that with the other command types in unit tests, but I have opened a PR with an E2E test that covers other cases: https://github.com/intel/llvm-test-suite/pull/79